op_crates/crypto/01_crypto.js

```diff
@@ -1,25 +1,206 @@
+        arrayBufferView instanceof Uint8ClampedArray
+      )
+    ) {
+      throw new DOMException(
+        "The provided ArrayBufferView is not an integer array type",
+        "TypeMismatchError",
+      );
+    }
+    if (arrayBufferView.byteLength > 65536) {
+      throw new DOMException(
+        `The ArrayBufferView's byte length (${arrayBufferView.byteLength}) exceeds the number of bytes of entropy available via this API (65536)`,
+        "QuotaExceededError",
+      );
     }
     const ui8 = new Uint8Array(
-      typedArray.buffer,
-      typedArray.byteOffset,
-      typedArray.byteLength,
+      arrayBufferView.buffer,
+      arrayBufferView.byteOffset,
+      arrayBufferView.byteLength,
     );
-    core.jsonOpSync("op_get_random_values", {}, ui8);
-    return typedArray;
+    core.jsonOpSync("op_crypto_get_random_values", {}, ui8);
+    return arrayBufferView;
   }
+
+  // Algorithm normalization, which involves storing the expected data for
+  // each pair of algorithm and crypto operation, is done on the JS side because
+  // it is convenient.
+  // Shall it stay that way, or are we better to move it on the Rust side?
+
+  // We need this method after initialization, anytime we need to normalize
+  // a provided algorithm. We store it here to prevent prototype pollution.
+  const toUpperCase = String.prototype.toUpperCase;
+
+  class RegisteredAlgorithmsContainer {
+    #nameIndex;
+    #definitions;
+
+    constructor(definitions) {
+      this.#nameIndex = Object.create(null);
+      this.#definitions = Object.create(null);
+      for (const [name, definition] of Object.entries(definitions)) {
+        this.#nameIndex[name.toUpperCase()] = name;
+        this.#definitions[name] = definition;
+      }
+    }
+
+    getDefinition(name) {
+      return this.#definitions[name];
+    }
+
+    normalizeName(providedName) {
+      const upperCaseName = toUpperCase.call(providedName);
+      return this.#nameIndex[upperCaseName];
+    }
+  }
+
+  const supportedAlgorithms = {
+    digest: new RegisteredAlgorithmsContainer({
+      "SHA-1": {},
+      "SHA-256": {},
+      "SHA-384": {},
+      "SHA-512": {},
+    }),
+  };
+
+  const digestByteLengths = {
+    "SHA-1": 20,
+    "SHA-256": 32,
+    "SHA-384": 48,
+    "SHA-512": 64,
+  };
+
+  function normalizeAlgorithm(algorithm, registeredAlgorithms) {
+    let alg;
+    if (typeof algorithm === "string") {
+      alg = { name: algorithm };
+    }
+    if (typeof algorithm === "object" && algorithm !== null) {
+      if (typeof algorithm.name !== "string") {
+        throw new TypeError("Algorithm name is missing or not a string");
+      }
+      const algorithmName = registeredAlgorithms.normalizeName(algorithm.name);
+      if (algorithmName === undefined) {
+        throw new DOMException(
+          "Unrecognized algorithm name",
+          "NotSupportedError",
+        );
+      }
+      const definition = registeredAlgorithms.getDefinition(algorithmName);
```

Comment:
Can be combined into a single step, right? `getDefinition()` normalizes the name before lookup, and you throw `DOMException` when the return value is undefined.

Reply:
That's how it was originally, but the spec mentions that the name must be the normalised one, not the original one.
So, I need to get both the normalised name _and_ the definition.
(Also, `getDefinition()` does not normalise the name before lookup 🤔)